### PR TITLE
[sdl2-mixer] Update features and fix static linking

### LINKED ISSUE #10805: [sdl2-mixer:x64-uwp] build failure
**Host Environment**
- OS: Microsoft Windows 10 Pro 1909

- Compiler:  MSVC 19.26.28720.3

**To Reproduce**
`PS C:\vcpkg> .\vcpkg install sdl2-mixer:x64-uwp`

**Failure logs** 
PS C:\vcpkg> .\vcpkg install `sdl2-mixer:x64-uwp`
Computing installation plan...
The following packages will be built and installed:
    sdl2-mixer[core]:x64-uwp
Starting package 1/1: sdl2-mixer:x64-uwp
Building package sdl2-mixer[core]:x64-uwp...
-- Using cached C:/vcpkg/downloads/SDL2_mixer-2.0.4.zip
-- Using source at C:/vcpkg/buildtrees/sdl2-mixer/src/2.0.4-a895b09396
-- Configuring x64-uwp
-- Building x64-uwp-dbg
CMake Error at scripts/cmake/vcpkg_execute_build_process.cmake:136 (message):
    Command failed: C:/vcpkg/downloads/tools/cmake-3.14.0-windows/cmake-3.14.0-win32-x86/bin/cmake.exe --build
 . --config Debug --target install -- /p:VCPkgLocalAppDataDisabled=true /p:UseIntelMKL=No /m
    Working Directory: C:/vcpkg/buildtrees/sdl2-mixer/x64-uwp-dbg
    See logs for more information:
      C:\vcpkg\buildtrees\sdl2-mixer\install-x64-uwp-dbg-out.log

Call Stack (most recent call first):
  scripts/cmake/vcpkg_build_cmake.cmake:91 (vcpkg_execute_build_process)
  scripts/cmake/vcpkg_install_cmake.cmake:24 (vcpkg_build_cmake)
  ports/sdl2-mixer/portfile.cmake:49 (vcpkg_install_cmake)
  scripts/ports.cmake:90 (include)

Error: Building package sdl2-mixer:x64-uwp failed with: BUILD_FAILED
Please ensure you're using the latest portfiles with `.\vcpkg update`, then
submit an issue at https://github.com/Microsoft/vcpkg/issues including:
  Package: sdl2-mixer:x64-uwp
  Vcpkg version: 2020.02.04-nohash

Additionally, attach any relevant sections from the log files above.

4>music.obj : error LNK2019: unresolved external symbol native_midi_detect referenced in function open_music_type [C:\vcpkg\buildtrees\sdl2-mixer\x64-uwp-dbg\SDL2_mixer.vcxproj]

**Additional context**
I'm really noob, I may missing a lib in my environment, but I didn't have any problem building sdl2-mixer:x64-windows port...
 sorry for disturbing you guys!

### PR DESCRIPTION
**Pull request by FtZPetruska**
- [X] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)
- [X] SHA512s are updated for each updated download
- [X] The "supports" clause reflects platforms that may be fixed by this new version
- [X] Any fixed [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt) entries are removed from that file.
- [X] Any patches that are no longer applied are deleted from the port's directory.
- [X] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [X] Only one version is added to each modified port's versions file.

---

Follow-up PR to the discussions in [#29487](https://github.com/microsoft/vcpkg/pull/29487#discussion_r1103274302). (cc @dg0yt)

The first issue is about the backend selection and the impact of features.

The `fluidsynth`, `libflac`, `mpg123`, `nativemidi`, and `libvorbis` features of `sdl2-mixer` worked as alternatives instead of adding functionalities. This is due to SDL_mixer embedding additional backends for these formats through header-only libraries and in-tree source code. This leads to several issues:
- Embedding code with different licenses: Timidity is enabled by default, which is licensed under `(1) the GNU GPL, (2) the GNU LGPL, or (3) the Perl Artistic License` which may be an issue since SDL_mixer is only listed as Zlib.
- Providing unexpected backends with varying quality: with the current behaviour, a user can initialise and play back flac, MIDI, mp3 and ogg with no features built.
- Wrong backend selected internally: while the `nativemidi` feature was selected by default, it would always get shadowed by Timidity as the latter has a higher priority.

To address this issue, the following changes were made:
- When possible, disable embedded backends. (`dr_mp3`, `dr_flac`, `timidity`)
- Since the vorbis backend cannot be disabled, change `libvorbis` from a feature to a dependency.
- Remove the `nativemidi` feature and disable its backend. Outside of Windows and Apple platforms the option would be silently ignored, moreover, it seemed to be breaking UWP builds (#10805)

With the changes, I was able to confirm that every format only becomes available when the associated feature is selected.

The second issue is with consuming the dependencies and forwarding their usage requirements.

SDL_mixer uses custom Find modules in its CMakeLists and its installed Config to find its dependencies. This works mostly for dynamic builds but causes issues with static builds as the find modules do not know anything about transitive dependencies. Moreover, the target names used in the Find would sometimes differ from what the upstream library uses.

Since all packages with transitive dependencies used by `sdl2-mixer` now export an official CMake Config. The calls to `find_package`, `find_dependency`, and `target_link_libraries` have been patched to use the correct target names. This removes the need to patch the find modules, and explicitly find and link `Ogg` and `Opus` when the `opusfile` feature is selected. But most importantly, since the usage requirements are properly handled, this allows linking SDL_mixer statically without having to manually set `<lib>_LINK_LIBRARIES` for every dependency that was built.

A drawback of this patch is that selecting the `fluidsynth` feature now adds a dependency to pkg-config in order for the package to find its own dependencies.

---

This PR was tested with all features on:
- x64-windows
- x64-windows-static
- arm64-osx
- x64-linux

Also tested on:
- x64-uwp (with `libflac`, `libmodplug`, and `mpg123`)
- arm-uwp (with `libflac` and `mpg213`)

Fixes #10805.

### CODE REVIEW

File: ports/sdl2-mixer/vcpkg.json [hunk 1]
```diff
@@ -26,34 +25,25 @@
...
       "dependencies": [
         "libmodplug"
       ]
     },
-    "libvorbis": {
-      "description": "Use libvorbis to play OGG Vorbis audio format instead of the built-in stb_vorbis.",
-      "dependencies": [
-        "libvorbis"
-      ]
-    },
```
Comment (dg0yt) [inline ports/sdl2-mixer/vcpkg.json:45]: We should keep the feature as a no-op in order to not break upgrades of existing installations. Possible description: "The libvorbis backend is always enabled. This feature is kept for compatibility."

(vcpkg tool will be more tolerant soon, but we don't know about downstream references to this feature, including manifests in user projects.)

**Review by dg0yt — COMMENTED** [review]
This PR looks good to me. Thanks for the extend information.

**Comment by oleg-derevenetz** [comment]
Hi all,

> Providing unexpected backends with varying quality: with the current behaviour, a user can initialise and play back flac, MIDI, mp3 and ogg with no features built.

How is this an issue? If I prefer to use a built-in backend instead of an external dependency (that comes with additional DLLs), I should be able to do this.

> Wrong backend selected internally: while the `nativemidi` feature was selected by default, it would always get shadowed by Timidity as the latter has a higher priority.

Only if Timidity is able to play this MIDI (i.e. has the necessary config file & GUS patches). Otherwise the `nativemidi` will be used. The same is valid for the Fluidsynth backend.

> When possible, disable embedded backends. (`dr_mp3`, `dr_flac`, `timidity`)

If there is no way to enable them back, this is plain wrong.

> Remove the `nativemidi` feature and disable its backend.

Why? What if I want to use the `nativemidi` backend?

It looks like a bunch of drastic changes without taking into account the user's point of view.

**Comment by dg0yt** [comment]
Let us take the time to understand the motivation and the implications. Such PRs often benefit from feedback.

> > When possible, disable embedded backends. (`dr_mp3`, `dr_flac`, `timidity`)
> 
> If there is no way to enable them back, this is plain wrong.

How can you currently enable `dr_mp3` once one package starts to depend on feature `mpg123`?

AFAIU this an example of the key problem with the current state of the port.
- Features should control features (mp3 support on/off), not dependencies (lib mpg123).
  It goes without saying that a feature may come with some dependencies.
- Features shall not control alternative APIs.
  It takes some insight to see if this is the case. (And I admit I don't have enough insight for sdl2-mixer now.)
  For choosing alternatives, there are (limited) other customization points.

> What if I want to use the nativemidi backend?

Given that there this a CMake port, there is already one option: Use an overlay triplet with this extra code:
~~~cmake
if(PORT STREQUAL "sdl2-mixer")
    list APPEND(VCPKG_CMAKE_CONFIGURE_OPTIONS "-DSDL2MIXER_MIDI_NATIVE=ON")
endif()
~~~

**Comment by oleg-derevenetz** [comment]
Hi @dg0yt

> How can you currently enable `dr_mp3` once one package starts to depend on feature `mpg123`?

Currently, I just will not build it with `mpg123` feature. With this PR, I should be either build it with additional dependencies or has no support for the corresponding formats at all. What if I just want to use the built-in backend? And why disable it at all, since it doesn't bring other dependencies? Who and how will suffer from this?

> AFAIU this an example of the key problem with the current state of the port.

While the policy matters, changes still shouldn't break things. There still should be a way to enable the built-in backends if they are disabled by default.

> Given that there this a CMake port, there is already one option: Use an overlay triplet with this extra code:

Yes, I can make "patches" like this. But is this really a user-friendly way?

**Comment by dg0yt** [comment]
> Currently, I just will not build it with mpg123 feature.

You loose this control when you start to use another port which requests this feature.

But no need to repeat your points. Give some time for taking them into account.

**Comment by FtZPetruska** [comment]
Thank you for taking time to leave your feedback!

I understand your point about taking additional external dependencies. But I would like to nuance this with a practical use case that happened to me.

I was working on a small program that used SDL_mixer for OGG playback. It was running on:
- Linux, with the dependencies provided by the system package manager 
- macOS, with the dependencies provided by the Homebrew package manager
- Windows, with the dependencies provided by vcpkg

On Linux and macOS, the program would work without any issue, both the system package manager and Homebrew disable the header-only backends in favour of using external libraries (mpg123, libvorbis, etc)

On Windows, I would get a very odd behaviour:
- In debug builds, the program would terminate with a segmentation fault on shutdown
- In release builds, the program would terminate with a segmentation fault as soon as the music was halted and changed.

What it came down to was the use of the backend for OGG playback. On Linux and macOS libvorbis was used and worked fine, whereas on Windows it would use `stb_vorbis` out of the box. The latter would corrupt the audio thread spawned by SDL.

The main takeaway is that having different backends each with varying limitation for the same audio format may lead to troublesome issues. In the case of the issue I described above, it took several hours to pinpoint the origin down to the backend, but it could be more difficult to diagnostic in a larger codebase that involves multiple people where one could have a feature installed from another project.


On that note, as mentioned previously, if a user wants a custom build of SDL_mixer with header-only backends, with nativemidi or timidity, or even without WAV support, they can absolutely do that by passing any option they need through an overlay triplet.

Thank you for your understanding!

**Comment by oleg-derevenetz** [comment]
Hi @FtZPetruska

> On Linux and macOS libvorbis was used and worked fine, whereas on Windows it would use `stb_vorbis` out of the box. The latter would corrupt the audio thread spawned by SDL.

I'm not sure what the library package maintainers' approach should look like "I got a crash with this backend once, so this is not the backend you are looking for" :) I have always believed that developers (especially qualified ones) should be given a choice. Although, as you correctly said...

> if a user wants a custom build of SDL_mixer with header-only backends, with nativemidi or timidity, or even without WAV support, they can absolutely do that by passing any option they need through an overlay triplet.

Just curious what would happen if you got a crash in `libvorbis` instead of `stb_vorbis` :) Perhaps then you would have decided to ban all external dependencies in favor of built-in ones :)

**Comment by FtZPetruska** [comment]
Thank you for your comment.

That is a fair point and I will take it into consideration.

**Comment by MonicaLiu0311** [comment]
usage:
```
sdl2-mixer provides CMake targets:

    find_package(SDL2_mixer CONFIG REQUIRED)
    target_link_libraries(main PRIVATE $<IF:$<TARGET_EXISTS:SDL2_mixer::SDL2_mixer>,SDL2_mixer::SDL2_mixer,SDL2_mixer::SDL2_mixer-static>)
```

When testing usage, the following error occurs:
```
>------ Build All started: Project: CMakeFindUsage, Configuration: x64-Release ------
  MSBuild version 17.4.1+9a89d02ff for .NET Framework
    Checking Build System
    Building Custom Rule C:/Users/monica/source/repos/CMakeFindUsage/CMakeFindUsage/CMakeLists.txt
    CMakeFindUsage.cpp
C:\Users\monica\source\repos\CMakeFindUsage\out\build\x64-Release\CMakeFindUsage\MSVCRT.lib(exe_main.obj) : error LNK2019: unresolved external symbol main referenced in function "int __cdecl __scrt_common_main_seh(void)" (?__scrt_common_main_seh@@YAHXZ) 
C:\Users\monica\source\repos\CMakeFindUsage\out\build\x64-Release\CMakeFindUsage\Release\CMakeFindUsage.exe : fatal error LNK1120: 1 unresolved externals 

Build All failed.
```

<details>
<summary><b>CMakeFindUsage.cpp</b></summary>
<pre>
#include &#60;iostream&#62;
#include "SDL.h"

using namespace std;

int main()
{
        cout << "Hello CMake." << endl;
        return 0;
}
</pre></details>


<details>
<summary><b>CMakeLists.txt</b></summary>
<pre>
cmake_minimum_required (VERSION 3.8)

project ("CMakeFindUsage")

find_package(SDL2_mixer CONFIG REQUIRED)

add_executable (CMakeFindUsage "CMakeFindUsage.cpp")

target_link_libraries(CMakeFindUsage PRIVATE *$<IF:$<TARGET_EXISTS:SDL2_mixer::SDL2_mixer>,SDL2_mixer::SDL2_mixer,SDL2_mixer::SDL2_mixer-static>*)
</pre></details>


<details>
<summary><b>CMakeSettings.json</b></summary>
<pre>
{
  "configurations": [
    {
      "name": "x64-Release",
      "generator": "Visual Studio 17 2022 Win64",
      "configurationType": "Release",
      "inheritEnvironments": [ "msvc_x64" ],
      "buildRoot": "${projectDir}\\out\\build\\${name}",
      "installRoot": "${projectDir}\\out\\install\\${name}",
      "cmakeCommandArgs": "-DCMAKE_TOOLCHAIN_FILE=E:/sdl2-mixer/scripts/buildsystems/vcpkg.cmake",
      "buildCommandArgs": "",
      "ctestCommandArgs": "",
      "variables": [
        {
          "name": "VCPKG_TARGET_TRIPLET",
          "value": "x64-windows-static"
        }
      ]
    }
  ]
}
</pre></details>

**Comment by dg0yt** [comment]
> CMakeFindUsage.cpp
> 
> #include "SDL.h

@JonLiu1993  This header doesn't belong to port sdl2-mixer.

**Comment by MonicaLiu0311** [comment]
The usage test passed (header files found):
```
sdl2-mixer provides CMake targets:

    find_package(SDL2_mixer CONFIG REQUIRED)
    target_link_libraries(main PRIVATE $<IF:$<TARGET_EXISTS:SDL2_mixer::SDL2_mixer>,SDL2_mixer::SDL2_mixer,SDL2_mixer::SDL2_mixer-static>)
```

All features are tested successfully in the following triplet:
```
x86-windows
x64-windows
x64-windows-static
```

**Comment by dan-shaw** [comment]
Do we know which backend supports the most MIDI configurations (Timidity, nativemidi, fluidsynth)? This could help us in determining what makes sense as the default. Also, do we know why other system package managers disable the header only backends in favor of external libraries?

**Comment by FtZPetruska** [comment]
> Do we know which backend supports the most MIDI configurations (Timidity, nativemidi, fluidsynth)? This could help us in determining what makes sense as the default.

As far as I know:
- nativemidi supports only Apple and Windows (non-UWP).
- fluidsynth supports Linux, Apple, and Windows (non-UWP, because of glib2 dependency)
- timidity support Linux, Apple, and Windows but according to SDL2_mixer's README:
  ```
  To play MIDI files using Timidity, you'll need to get a complete set of GUS patches from:
  http://www.libsdl.org/projects/mixer/timidity/timidity.tar.gz
  and unpack them in /usr/local/lib under UNIX, and C:\ under Win32.
  ```

> Also, do we know why other system package managers disable the header only backends in favor of external libraries?

Here are [Debian's stance](https://www.debian.org/doc/debian-policy/ch-source.html#s-embeddedfiles) ([see also footnote 18](https://www.debian.org/doc/debian-policy/ch-source.html#id37), followed by Ubuntu), [Fedora's stance](https://docs.fedoraproject.org/en-US/packaging-guidelines/#bundling), and [Homebrew's stance](https://docs.brew.sh/Acceptable-Formulae#stuff-that-requires-vendored-versions-of-homebrew-formulae).

Their key points are:
- embedding libraries may lead to a security risk;
  If a security vulnerability is found in a header-only vulnerability, we are depending on each project that uses it to update their copy in their codebase. Whereas, if a vulnerability is found within a library dependency, only it needs to be updated.
- if the package manager already provides the feature as an external library, it should depend on the external library instead of embedding the code.
  In the case of sdl2-mixer, every embedded library has an alternative external library available in vcpkg.

Similarly, [vcpkg's stance](https://learn.microsoft.com/en-us/vcpkg/contributing/maintainer-guide#do-not-use-vendored-dependencies) on vendored libraries is similar to other package managers.

I hope this addresses your questions!

**Comment by MonicaLiu0311** [comment]
@dan-shaw

File: ports/sdl2-mixer/vcpkg.json [hunk 2]
```diff
@@ -26,34 +25,28 @@
...
       "dependencies": [
         "libmodplug"
       ]
     },
     "libvorbis": {
-      "description": "Use libvorbis to play OGG Vorbis audio format instead of the built-in stb_vorbis.",
-      "dependencies": [
-        "libvorbis"
-      ]
+      "description": "The libvorbis backend is always enabled. This feature is kept for compatibility."
```
Comment (BillyONeal) [inline ports/sdl2-mixer/vcpkg.json:40]: I think we should just get rid of the feature then.

Comment (FtZPetruska) [reply ports/sdl2-mixer/vcpkg.json]: Thank you for your comment. When the PR was initially opened, this change had not been released, glad to see it's out!

I removed the feature in [210059b](https://github.com/microsoft/vcpkg/pull/29681/commits/210059b8851126efb64d0b66839467a5055f8363).

**Pull request merged**